---
title: "1_alignment_and_counting"
author: "Samuel Palframan"
date: "2024-02-17"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# RNA-seq Analysis of Murine Mammary data

## Alignment and counting of RNA-seq data

Please note that this code was derived from a COMBINE training workshop.
Original Authors: Maria Doyle, Belinda Phipson, Matt Ritchie, Anna Trigos, Harriet Dashnow, Charity Law.
Source: `https://bioinformatics-core-shared-training.github.io/RNAseq-R/`

## OPTIONAL: Create R DataFrame from Fastq file using `baseq` for EDA
```{r}
# Install and/or load `baseq` to read in FASTQ file as R DataFrame.
#BiocManager::install("baseq")
library(baseq)

# Set working directory
setwd("~/Jupyter Notebook/Module_X_R/RNA-Seq_Murine_Mammary_COMBINE")

# Specify file path
file_path <- "./data/SRR1552444.fastq.gz"

# Read in Fastq file as R DataFrame
read.fastq_to_df(file_path)

# Create Vector of Header column
header <- SRR1552444.fastq.gz[,"Header"]
print(head(header))
```

## Load Rsubread
```{r}
# Install and/or load `Rsubread`.
#BiocManager::install("Rsubread")
library(Rsubread)
```

## Load all FASTQ files from `/.data` folder using `list.files` command
```{r}
fastq.files <- list.files(path = "./data", pattern = ".fastq.gz$", full.names = TRUE)
fastq.files
```

## Alignment
### Build the index
```{r}
# Note that Mouse Chromosome 1 .fa file (chr1.fa) was dowloaded from: http://hgdownload.soe.ucsc.edu/goldenPath/mm39/chromosomes/
# The code below was not run in this case because the reference chromosome 1 index was supplied
#buildindex(basename="chr1_mm10",reference="./data/chr1.fa")
```

### Align reads to chromosome 1 of reference genome
```{r}
# Use `align` command to align reads from the FASTQ files
align(index="data/chr1_mm10",readfile1=fastq.files)

# To see how many parameters can be changed, use `args` function
args(align)

# To open Help information for `align`
?align
```

### Use `propmapped` function to generate summary of proportion of reads mapped to reference genome
```{r}
bam.files <- list.files(path = "./data", pattern = ".BAM$", full.names = TRUE)
bam.files

# `propmapped` shows that approx. 12 % of genes mapped to Mouse Chromosome 1
props <- propmapped(files=bam.files)
props
```

## Quality Control
```{r}
# Extract the quality scores of 100 reads from the file "SRR1552450.fastq.gz"
qs <- qualityScores(filename="data/SRR1552450.fastq.gz",nreads=100)

# Check class or type of variable created
class(qs)

# Check how many scores are below 20 (i.e worse than 1 in 100 chance of being wrong)
# Note that a score above 20 is usually acceptable, but 30 and 40 are much better scores
poor_score <- sum(qs < 20)
print(poor_score)

# Check dimensions of qs
dimensions <- dim(qs)
print(dimensions)

# Check first few elements of qs with head
head(qs)

# Note that a quality score of 30 corresponds to a 1 in 1000 chance of an incorrect base call. 
# A quality score of 10 = 1 in 10, 20 = 1 in 100, 30 = 1 in 1000, 40 = 1 in 10,000 etc. chance of an incorrect base call.

# Create Boxplot of qs to visualise Phred quality scores
boxplot(qs)
```

## Counting
```{r}
# Use `featureCounts` function to count mapped reads to mouse genes
# Each sample is a separate column, each gene is a row
fc <- featureCounts(bam.files, annot.inbuilt = "mm10")

# Check names stored in fc
names(fc)

# Check featureCounts stats
fc$stat

# Check the gene counts for each sample
fc$counts

# Check only the first 6 lines of each file
head(fc$counts)

# Check dimensions to view number of genes (i.e 27,179 genes in 12 files)
dim(fc$counts)

# Check annotation information from featureCounts
head(fc$annotation)
```

### EXTRA: Create DataFrame of gene counts for EDA and potential .csv export
```{r}
# Create DataFrame of gene counts
fc_df <- as.data.frame(fc$counts)

# Check whether there are NAN values
any(is.na(fc_df))

# Check Structure of fc_df
str(fc_df)

# Check and create vector of column names of fc_df
column_names <- as.vector(colnames(fc_df))
print(column_names)

# Sort fc_df in descending order by desired column/experiment/file
sorted_df <- fc_df[order(-fc_df[, column_names[1]]), ]
```